docs/proposals/service-catalog.md

```diff
@@ -0,0 +1,143 @@
+
+## Claiming a service
+
+When you want to use an entry from the service catalog, you create a `ServiceClaim` that references the desired entry. A controller processes new claims for admission. This determines if the user who created the claim is allowed to consume the entry from the catalog. This decision can be flexible: it could be automated based on policy, or it could support manual intervention and workflow.
+
+Once the claim has been admitted, a controller performs the provisioning process based on the service catalog entry's type.
+
+For a reference claim, a new service is created in the user's namespace. The new service somehow needs to be a CNAME to the source. Additionally, for each configuration resource referenced by the source service, the controller creates a clone of it in the user's namespace.
+
+For a template claim, the template is processed and its items are created in the user's namespace (see [OpenShift templates](https://docs.openshift.org/latest/dev_guide/templates.html) for more details on templates).
+
+For a provision claim, the controller sends requests to the specified "service broker" endpoint to perform whatever provisioning actions the broker is coded to do.
+
+Additional fulfillment types are possible as long as there is a controller that is handling them.
+
+TODO poking holes in cluster firewall for cross-namespace connections
+
+# Linking services
+
+Claiming a service catalog entry only creates resources in the user's namespace. If all you need is a service and its DNS entry, this may be sufficient for your pods to function. But if you need the configuration data injected into your pod, it would be nice to make that easier to do.
+
+We want to add the ability to link a service to a deployable resource such as a Deployment. This could look something like:
+
+    kubectl link postgresql deployment/web
+    service "postgresql" linked to deployemnt "web"
+    configdata "postgresql-options" linked to deployment "web" as a volume
+    secret "postgresql-credentials" linked to deployment "web" as a volume
+
+This command would automatically inject the ConfigData and Secret objects as volumes into the Deployment. This could be flexible as well, allowing you instead to expose these items as environment variables.
+
+In the example above, the volumes could potentially be mounted as:
+
+/var/run/kubernetes.io/links/configdata/postgresql-options
+/var/run/kubernetes.io/links/secrets/postgresql-credentials
+
+
+# Cross-namespace networking
+
+If the cluster has multi-tenant network isolation enabled, then a pod in namespace A won't be able to talk to a service in namespace B. We should look into ways to automatically manipulate the isolation rules to "poke holes" when claims are provisioned, to make the desired connectivity work, and to "unpoke" when the connectivity is no longer needed.
+
+# TODOs
+
+- One catalog or multiple?
```

Comment:
I could easily imagine different orgs offering separate catalogs, in addition to a cluster-wide one.